File: tests/requests/recurring_spec.rs
use chrono::Duration;
use loco_rs::testing::prelude::*;
use serial_test::serial;

use bookkeeping_backend::api::time;
use bookkeeping_backend::app::App;

use super::prepare_data::{
    auth_header, count_occurrences, count_occurrences_without_transaction, count_transactions,
    create_rule, find_rule, json_body, list_transactions, register_and_login,
};

fn in_hours(hours: i64) -> String {
    time::format_datetime(&(time::now_local() + Duration::hours(hours)))
}

#[tokio::test]
#[serial]
async fn creates_and_pauses_resumes_a_rule() {
    request::<App, _, _>(|request, ctx| async move {
        let fixture = register_and_login(&request, &ctx).await;
        let (name, value) = auth_header(&fixture.token);
        let created = request
            .post("/api/v1/recurring_rules")
            .add_header(name.clone(), value.clone())
            .json(&serde_json::json!({
                "account_id": fixture.account_id,
                "kind": "expense",
                "amount_cents": 1000,
                "frequency": "daily",
                "interval": 1,
                "start_on": time::today().to_string(),
                "next_run_at": in_hours(1)
            }))
            .await;
        assert_eq!(created.status_code(), 201);
        let id = json_body(&created.text())["data"]["id"]
            .as_str()
            .unwrap()
            .to_string();

        let paused = request
            .post(&format!("/api/v1/recurring_rules/{id}/pause"))
            .add_header(name.clone(), value.clone())
            .await;
        assert_eq!(json_body(&paused.text())["data"]["status"], "paused");

        let resumed = request
            .post(&format!("/api/v1/recurring_rules/{id}/resume"))
            .add_header(name, value)
            .await;
        assert_eq!(json_body(&resumed.text())["data"]["status"], "active");
    })
    .await;
}

#[tokio::test]
#[serial]
async fn updates_a_rule_and_clears_nullable_fields() {
    request::<App, _, _>(|request, ctx| async move {
        let fixture = register_and_login(&request, &ctx).await;
        let (name, value) = auth_header(&fixture.token);
        let created = request
            .post("/api/v1/recurring_rules")
            .add_header(name.clone(), value.clone())
            .json(&serde_json::json!({
                "account_id": fixture.account_id,
                "category_id": fixture.expense_category_id,
                "kind": "expense",
                "amount_cents": 1000,
                "frequency": "daily",
                "interval": 1,
                "start_on": time::today().to_string(),
                "note": "before",
                "next_run_at": in_hours(1)
            }))
            .await;
        assert_eq!(created.status_code(), 201);
        let id = json_body(&created.text())["data"]["id"]
            .as_str()
            .unwrap()
            .to_string();

        // Partial update: changed fields apply, omitted fields stay.
        let updated = request
            .patch(&format!("/api/v1/recurring_rules/{id}"))
            .add_header(name.clone(), value.clone())
            .json(&serde_json::json!({ "amount_cents": 2500, "note": "after" }))
            .await;
        assert_eq!(updated.status_code(), 200);
        let body = json_body(&updated.text());
        assert_eq!(body["data"]["amount_cents"], 2500);
        assert_eq!(body["data"]["note"], "after");
        assert_eq!(body["data"]["interval"], 1);
        assert_eq!(body["data"]["frequency"], "daily");

        // Explicit null clears the nullable category; omitting it would not.
        let cleared = request
            .patch(&format!("/api/v1/recurring_rules/{id}"))
            .add_header(name, value)
            .json(&serde_json::json!({ "category_id": null }))
            .await;
        assert_eq!(cleared.status_code(), 200);
        assert!(json_body(&cleared.text())["data"]["category_id"].is_null());
    })
    .await;
}

#[tokio::test]
#[serial]
async fn rejects_invalid_rule_updates() {
    request::<App, _, _>(|request, ctx| async move {
        let fixture = register_and_login(&request, &ctx).await;
        let (name, value) = auth_header(&fixture.token);
        let rule = create_rule(
            &ctx.db,
            &fixture.user_id,
            &fixture.account_id,
            0,
            time::now_local() + Duration::hours(1),
            None,
            None,
        )
        .await;

        let response = request
            .patch(&format!("/api/v1/recurring_rules/{}", rule.id))
            .add_header(name, value)
            .json(&serde_json::json!({ "amount_cents": 0, "interval": 0 }))
            .await;
        assert_eq!(response.status_code(), 422);
        let body = json_body(&response.text());
        assert_eq!(body["error"]["code"], "validation_error");
    })
    .await;
}

#[tokio::test]
#[serial]
async fn lists_and_filters_rules_by_status() {
    request::<App, _, _>(|request, ctx| async move {
        let fixture = register_and_login(&request, &ctx).await;
        create_rule(
            &ctx.db,
            &fixture.user_id,
            &fixture.account_id,
            0,
            time::now_local() + Duration::hours(1),
            None,
            Some("active rule".to_string()),
        )
        .await;
        create_rule(
            &ctx.db,
            &fixture.user_id,
            &fixture.account_id,
            1,
            time::now_local() + Duration::hours(1),
            None,
            Some("paused rule".to_string()),
        )
        .await;
        let (name, value) = auth_header(&fixture.token);

        let all = request
            .get("/api/v1/recurring_rules")
            .add_header(name.clone(), value.clone())
            .await;
        let notes: Vec<String> = json_body(&all.text())["data"]
            .as_array()
            .unwrap()
            .iter()
            .map(|row| row["note"].as_str().unwrap().to_string())
            .collect();
        assert!(notes.contains(&"active rule".to_string()));
        assert!(notes.contains(&"paused rule".to_string()));

        let paused = request
            .get("/api/v1/recurring_rules?status=paused")
            .add_header(name.clone(), value.clone())
            .await;
        let rows = json_body(&paused.text());
        let rows = rows["data"].as_array().unwrap();
        assert_eq!(rows.len(), 1);
        assert_eq!(rows[0]["note"], "paused rule");

        let ended = request
            .get("/api/v1/recurring_rules?status=ended")
            .add_header(name.clone(), value.clone())
            .await;
        assert!(json_body(&ended.text())["data"]
            .as_array()
            .unwrap()
            .is_empty());

        let bogus = request
            .get("/api/v1/recurring_rules?status=bogus")
            .add_header(name, value)
            .await;
        assert_eq!(bogus.status_code(), 200);
        assert!(json_body(&bogus.text())["data"]
            .as_array()
            .unwrap()
            .is_empty());
    })
    .await;
}

#[tokio::test]
#[serial]
async fn run_now_materializes_once_then_conflicts() {
    request::<App, _, _>(|request, ctx| async move {
        let fixture = register_and_login(&request, &ctx).await;
        let (name, value) = auth_header(&fixture.token);
        let created = request
            .post("/api/v1/recurring_rules")
            .add_header(name.clone(), value.clone())
            .json(&serde_json::json!({
                "account_id": fixture.account_id,
                "kind": "expense",
                "amount_cents": 1000,
                "frequency": "daily",
                "interval": 1,
                "start_on": time::today().to_string(),
                "next_run_at": in_hours(1)
            }))
            .await;
        let id = json_body(&created.text())["data"]["id"]
            .as_str()
            .unwrap()
            .to_string();

        let first = request
            .post(&format!("/api/v1/recurring_rules/{id}/run_now"))
            .add_header(name.clone(), value.clone())
            .await;
        assert_eq!(first.status_code(), 200);
        assert_eq!(json_body(&first.text())["data"]["net_amount_cents"], 1000);

        let second = request
            .post(&format!("/api/v1/recurring_rules/{id}/run_now"))
            .add_header(name, value)
            .await;
        assert_eq!(second.status_code(), 409);
        assert_eq!(
            json_body(&second.text())["error"]["code"],
            "already_materialized"
        );
    })
    .await;
}

#[tokio::test]
#[serial]
async fn catch_up_materializes_exactly_once() {
    request::<App, _, _>(|request, ctx| async move {
        let fixture = register_and_login(&request, &ctx).await;
        create_rule(
            &ctx.db,
            &fixture.user_id,
            &fixture.account_id,
            0,
            time::now_local() - Duration::days(2),
            None,
            None,
        )
        .await;

        let (name, value) = auth_header(&fixture.token);
        request
            .get("/api/v1/me")
            .add_header(name.clone(), value.clone())
            .await;
        let count = count_transactions(&ctx.db, &fixture.user_id, 1).await;
        assert!(count >= 1);

        request.get("/api/v1/me").add_header(name, value).await;
        assert_eq!(
            count_transactions(&ctx.db, &fixture.user_id, 1).await,
            count
        );
    })
    .await;
}

#[tokio::test]
#[serial]
async fn skip_next_does_not_create_a_transaction() {
    request::<App, _, _>(|request, ctx| async move {
        let fixture = register_and_login(&request, &ctx).await;
        let rule = create_rule(
            &ctx.db,
            &fixture.user_id,
            &fixture.account_id,
            0,
            time::now_local() + Duration::hours(1),
            None,
            None,
        )
        .await;

        let (name, value) = auth_header(&fixture.token);
        let response = request
            .post(&format!("/api/v1/recurring_rules/{}/skip_next", rule.id))
            .add_header(name, value)
            .await;
        assert_eq!(response.status_code(), 200);
        assert_eq!(
            count_occurrences_without_transaction(&ctx.db, &rule.id).await,
            1
        );
        assert_eq!(count_transactions(&ctx.db, &fixture.user_id, 1).await, 0);
    })
    .await;
}

#[tokio::test]
#[serial]
async fn does_not_recreate_a_deleted_recurring_transaction() {
    request::<App, _, _>(|request, ctx| async move {
        let fixture = register_and_login(&request, &ctx).await;
        let rule = create_rule(
            &ctx.db,
            &fixture.user_id,
            &fixture.account_id,
            0,
            time::now_local() - Duration::hours(1),
            None,
            None,
        )
        .await;

        let (name, value) = auth_header(&fixture.token);
        request
            .get("/api/v1/me")
            .add_header(name.clone(), value.clone())
            .await;

        let transactions = list_transactions(&ctx.db, &fixture.user_id, 1).await;
        assert_eq!(transactions.len(), 1);
        let deleted = request
            .delete(&format!("/api/v1/transactions/{}", transactions[0].id))
            .add_header(name.clone(), value.clone())
            .await;
        assert_eq!(deleted.status_code(), 204);

        request.get("/api/v1/me").add_header(name, value).await;
        assert_eq!(count_transactions(&ctx.db, &fixture.user_id, 1).await, 0);
        assert!(count_occurrences_without_transaction(&ctx.db, &rule.id).await >= 1);
    })
    .await;
}

#[tokio::test]
#[serial]
async fn ends_a_rule_after_its_end_date() {
    request::<App, _, _>(|request, ctx| async move {
        let fixture = register_and_login(&request, &ctx).await;
        let rule = create_rule(
            &ctx.db,
            &fixture.user_id,
            &fixture.account_id,
            0,
            time::now_local() - Duration::days(2),
            Some(time::today() - Duration::days(1)),
            None,
        )
        .await;

        let (name, value) = auth_header(&fixture.token);
        request.get("/api/v1/me").add_header(name, value).await;
        assert_eq!(find_rule(&ctx.db, &rule.id).await.status, 2);
    })
    .await;
}

#[tokio::test]
#[serial]
async fn backfill_disabled_materializes_only_the_latest_occurrence() {
    request::<App, _, _>(|request, ctx| async move {
        let fixture = register_and_login(&request, &ctx).await;
        let rule = create_rule(
            &ctx.db,
            &fixture.user_id,
            &fixture.account_id,
            0,
            time::now_local() - Duration::days(3),
            None,
            None,
        )
        .await;

        let (name, value) = auth_header(&fixture.token);
        request.get("/api/v1/me").add_header(name, value).await;

        assert_eq!(count_occurrences(&ctx.db, &rule.id).await, 4);
        assert_eq!(count_transactions(&ctx.db, &fixture.user_id, 1).await, 1);
    })
    .await;
}

#[tokio::test]
#[serial]
async fn backfill_enabled_materializes_every_due_occurrence() {
    request::<App, _, _>(|request, ctx| async move {
        let fixture = register_and_login(&request, &ctx).await;
        let rule = create_rule(
            &ctx.db,
            &fixture.user_id,
            &fixture.account_id,
            0,
            time::now_local() - Duration::days(3),
            None,
            None,
        )
        .await;

        std::env::set_var("RECURRING_BACKFILL_ENABLED", "true");
        let (name, value) = auth_header(&fixture.token);
        request.get("/api/v1/me").add_header(name, value).await;
        std::env::remove_var("RECURRING_BACKFILL_ENABLED");

        assert_eq!(count_occurrences(&ctx.db, &rule.id).await, 4);
        assert_eq!(count_transactions(&ctx.db, &fixture.user_id, 1).await, 4);
    })
    .await;
}

File: test/recurring.spec.ts
import {INestApplication} from "@nestjs/common";
import {afterAll, beforeAll, beforeEach, describe, expect, it} from "vitest";

import * as time from "../src/common/time";
import {DataSource} from "typeorm";
import {createApp, http, HttpClient, resetDatabase} from "./helpers/app";
import {authHeader, countOccurrences, countOccurrencesWithoutTransaction, countTransactions, createRule, findRule, listTransactions, registerAndLogin} from "./helpers/fixtures";

function inHours(hours: number): string {
    return time.formatDatetime(time.addHours(time.nowLocal(), hours));
}

describe("recurring rules", () => {
    let app: INestApplication;
    let client: HttpClient;
    let dataSource: DataSource;

    beforeAll(async () => {
        app = await createApp();
        client = http(app);
        dataSource = app.get(DataSource);
    });

    beforeEach(async () => {
        await resetDatabase(dataSource);
    });

    afterAll(async () => {
        await app.close();
    });

    it("creates and pauses/resumes a rule", async () => {
        const fixture = await registerAndLogin(client);
        const headers = authHeader(fixture.token);
        const created = await client
            .post("/api/v1/recurring_rules")
            .set(headers)
            .send({
                account_id: fixture.accountId,
                kind: "expense",
                amount_cents: 1000,
                frequency: "daily",
                interval: 1,
                start_on: time.toDbDate(time.today()),
                next_run_at: inHours(1),
            });
        expect(created.status).toBe(201);
        const id = created.body.data.id as string;

        const paused = await client.post(`/api/v1/recurring_rules/${id}/pause`).set(headers);
        expect(paused.body.data.status).toBe("paused");

        const resumed = await client.post(`/api/v1/recurring_rules/${id}/resume`).set(headers);
        expect(resumed.body.data.status).toBe("active");
    });

    it("updates a rule and clears nullable fields", async () => {
        const fixture = await registerAndLogin(client);
        const headers = authHeader(fixture.token);
        const created = await client
            .post("/api/v1/recurring_rules")
            .set(headers)
            .send({
                account_id: fixture.accountId,
                category_id: fixture.expenseCategoryId,
                kind: "expense",
                amount_cents: 1000,
                frequency: "daily",
                interval: 1,
                start_on: time.toDbDate(time.today()),
                note: "before",
                next_run_at: inHours(1),
            });
        expect(created.status).toBe(201);
        const id = created.body.data.id as string;

        const updated = await client.patch(`/api/v1/recurring_rules/${id}`).set(headers).send({amount_cents: 2500, note: "after"});
        expect(updated.status).toBe(200);
        expect(updated.body.data.amount_cents).toBe(2500);
        expect(updated.body.data.note).toBe("after");
        expect(updated.body.data.interval).toBe(1);
        expect(updated.body.data.frequency).toBe("daily");

        const cleared = await client.patch(`/api/v1/recurring_rules/${id}`).set(headers).send({category_id: null});
        expect(cleared.status).toBe(200);
        expect(cleared.body.data.category_id).toBeNull();
    });

    it("rejects invalid rule updates", async () => {
        const fixture = await registerAndLogin(client);
        const rule = await createRule(dataSource, fixture.userId, fixture.accountId, 0, time.addHours(time.nowLocal(), 1), null, null);

        const response = await client.patch(`/api/v1/recurring_rules/${rule.id}`).set(authHeader(fixture.token)).send({amount_cents: 0, interval: 0});
        expect(response.status).toBe(422);
        expect(response.body.error.code).toBe("validation_error");
    });

    it("lists and filters rules by status", async () => {
        const fixture = await registerAndLogin(client);
        await createRule(dataSource, fixture.userId, fixture.accountId, 0, time.addHours(time.nowLocal(), 1), null, "active rule");
        await createRule(dataSource, fixture.userId, fixture.accountId, 1, time.addHours(time.nowLocal(), 1), null, "paused rule");
        const headers = authHeader(fixture.token);

        const all = await client.get("/api/v1/recurring_rules").set(headers);
        const notes = (all.body.data as Array<Record<string, unknown>>).map(row => row.note);
        expect(notes).toContain("active rule");
        expect(notes).toContain("paused rule");

        const paused = await client.get("/api/v1/recurring_rules?status=paused").set(headers);
        expect(paused.body.data).toHaveLength(1);
        expect(paused.body.data[0].note).toBe("paused rule");

        const ended = await client.get("/api/v1/recurring_rules?status=ended").set(headers);
        expect(ended.body.data).toEqual([]);

        const bogus = await client.get("/api/v1/recurring_rules?status=bogus").set(headers);
        expect(bogus.status).toBe(200);
        expect(bogus.body.data).toEqual([]);
    });

    it("run_now materializes once then conflicts", async () => {
        const fixture = await registerAndLogin(client);
        const headers = authHeader(fixture.token);
        const created = await client
            .post("/api/v1/recurring_rules")
            .set(headers)
            .send({
                account_id: fixture.accountId,
                kind: "expense",
                amount_cents: 1000,
                frequency: "daily",
                interval: 1,
                start_on: time.toDbDate(time.today()),
                next_run_at: inHours(1),
            });
        const id = created.body.data.id as string;

        const first = await client.post(`/api/v1/recurring_rules/${id}/run_now`).set(headers);
        expect(first.status).toBe(200);
        expect(first.body.data.net_amount_cents).toBe(1000);

        const second = await client.post(`/api/v1/recurring_rules/${id}/run_now`).set(headers);
        expect(second.status).toBe(409);
        expect(second.body.error.code).toBe("already_materialized");
    });

    it("catch up materializes exactly once", async () => {
        const fixture = await registerAndLogin(client);
        await createRule(dataSource, fixture.userId, fixture.accountId, 0, time.addDays(time.nowLocal(), -2), null, null);

        const headers = authHeader(fixture.token);
        await client.get("/api/v1/me").set(headers);
        const count = await countTransactions(dataSource, fixture.userId, 1);
        expect(count).toBeGreaterThanOrEqual(1);

        await client.get("/api/v1/me").set(headers);
        expect(await countTransactions(dataSource, fixture.userId, 1)).toBe(count);
    });

    it("skip_next does not create a transaction", async () => {
        const fixture = await registerAndLogin(client);
        const rule = await createRule(dataSource, fixture.userId, fixture.accountId, 0, time.addHours(time.nowLocal(), 1), null, null);

        const response = await client.post(`/api/v1/recurring_rules/${rule.id}/skip_next`).set(authHeader(fixture.token));
        expect(response.status).toBe(200);
        expect(await countOccurrencesWithoutTransaction(dataSource, rule.id)).toBe(1);
        expect(await countTransactions(dataSource, fixture.userId, 1)).toBe(0);
    });

    it("does not recreate a deleted recurring transaction", async () => {
        const fixture = await registerAndLogin(client);
        const rule = await createRule(dataSource, fixture.userId, fixture.accountId, 0, time.addHours(time.nowLocal(), -1), null, null);

        const headers = authHeader(fixture.token);
        await client.get("/api/v1/me").set(headers);

        const transactions = await listTransactions(dataSource, fixture.userId, 1);
        expect(transactions).toHaveLength(1);
        const deleted = await client.delete(`/api/v1/transactions/${transactions[0].id}`).set(headers);
        expect(deleted.status).toBe(204);

        await client.get("/api/v1/me").set(headers);
        expect(await countTransactions(dataSource, fixture.userId, 1)).toBe(0);
        expect(await countOccurrencesWithoutTransaction(dataSource, rule.id)).toBeGreaterThanOrEqual(1);
    });

    it("ends a rule after its end date", async () => {
        const fixture = await registerAndLogin(client);
        const rule = await createRule(dataSource, fixture.userId, fixture.accountId, 0, time.addDays(time.nowLocal(), -2), time.addDays(time.today(), -1), null);

        await client.get("/api/v1/me").set(authHeader(fixture.token));
        expect((await findRule(dataSource, rule.id)).status).toBe(2);
    });

    it("backfill disabled materializes only the latest occurrence", async () => {
        const fixture = await registerAndLogin(client);
        const rule = await createRule(dataSource, fixture.userId, fixture.accountId, 0, time.addDays(time.nowLocal(), -3), null, null);

        await client.get("/api/v1/me").set(authHeader(fixture.token));
        expect(await countOccurrences(dataSource, rule.id)).toBe(4);
        expect(await countTransactions(dataSource, fixture.userId, 1)).toBe(1);
    });

    it("backfill enabled materializes every due occurrence", async () => {
        const fixture = await registerAndLogin(client);
        const rule = await createRule(dataSource, fixture.userId, fixture.accountId, 0, time.addDays(time.nowLocal(), -3), null, null);

        process.env.RECURRING_BACKFILL_ENABLED = "true";
        try {
            await client.get("/api/v1/me").set(authHeader(fixture.token));
        } finally {
            delete process.env.RECURRING_BACKFILL_ENABLED;
        }

        expect(await countOccurrences(dataSource, rule.id)).toBe(4);
        expect(await countTransactions(dataSource, fixture.userId, 1)).toBe(4);
    });
});
